rpk: print how leaders and replicas are distributed across brokers

daisukebe opened the pull request:
rpk cluster partitions list and balancer-status now have a capability to show leader count and partition count per broker.

Examples:
```
# show the distribution for all topics
$ rpk cluster partitions list --all | head
LEADER DISTRIBUTION
===================
BROKER  LEADER-COUNT
0       15
1       15
2       15
3       15

REPLICA DISTRIBUTION
====================
BROKER  PARTITION-COUNT
0       40
1       40
2       40
3       40

LIST OF PARTITIONS
==================
NAMESPACE  TOPIC  PARTITION  LEADER-ID  REPLICA-CORE   DISABLED
```

```
# show the distribution for topic foo
 $ rpk cluster partitions list foo foo2 | head
LEADER DISTRIBUTION
===================
BROKER  LEADER-COUNT
0       12
1       13
2       13
3       12

REPLICA DISTRIBUTION
====================
BROKER  PARTITION-COUNT
0       37
1       38
2       38
3       37

LIST OF PARTITIONS
==================
NAMESPACE  TOPIC  PARTITION  LEADER-ID  REPLICA-CORE   DISABLED
```

```
$ rpk cluster partitions balancer-status
BALANCER STATUS
===============
Status:                      ready
Seconds Since Last Tick:     16
Current Reassignment Count:  0

REPLICA DISTRIBUTION
====================
BROKER  PARTITION-COUNT
0       40
1       40
2       40
3       40
```

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* rpk: print leader and replica distribution across brokers

daisukebe committed:
rpk: print leader and replica distribution across brokers

rpk cluster partitions list and balancer-status now
have a capability to show leader count and partition
count per broker.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--6d5178c9-5d27-4517-9d2f-a6cf0c12ba97-->
<!--TEST RESULTS-->
<!--e886ae4e-d8ac-4729-b604-49bf72a73508-->
<!--TEST RESULTS-->
<!--4bc52bbf-df7e-4f5e-9a75-49fcf0eeff39-->
<!--TEST RESULTS-->
<!--699031c9-7d5b-48bd-b64e-8764aa82172c-->
<!--TEST RESULTS-->
<!--0a681c76-761b-4a84-ae20-73728ee94fac-->
<!--TEST RESULTS-->
<!--2a5cd65f-3674-41e2-ae14-7202808b6e72-->
<!--TEST RESULTS-->
<!--ac990350-d5a3-40bd-a070-593bcc51be15-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62378">build#62378</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| distributed_kv_stm_tests_rpunit.distributed_kv_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62378#0195453a-1c9c-402f-8e1d-77e9cc401e3e | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_invalid_record_action_runtime_change.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_topic_property=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62378#01954580-1e32-4f17-b968-9a4c3da2a015 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62953">build#62953</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62953#01958852-73a8-4932-af58-900b089eb6bd | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63055">build#63055</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63055#01959063-116d-4a86-9d57-adaa6f6d3699 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.PartitionMovementTest.test_dynamic.num_to_upgrade=0 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63055#01959063-1169-4ef4-9b2b-f5899db55fb0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63353">build#63353</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63353#0195af38-7eba-4381-afb9-b338e61226be | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63353#0195af52-0f0e-42f3-a46e-259bd4c2d619 | FLAKY | 1/2 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_normalize.dataset_type=JSON | ducktape | https://buildkite.com/redpanda/redpanda/builds/63353#0195af52-0f0e-42f3-a46e-259bd4c2d619 | FLAKY | 1/7 |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_normalize.dataset_type=JSON | ducktape | https://buildkite.com/redpanda/redpanda/builds/63353#0195af38-7eb9-4202-a9c2-51d9bb4a9806 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63428">build#63428</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/63428#0195b392-7447-444e-bfc8-f5184ecb112f | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64007">build#64007</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_when_segment_size_smaller_than_chunk_size | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa3-696d-4327-8a56-c4f10fce770e | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa3-696b-42ef-89e9-2983e84523b7 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa3-696b-4cff-b8f5-042aacaf68ea | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa6-c42f-42cc-aa53-1b9b455cd9c0 | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa6-c42c-42ca-bd89-59398e0acd42 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move.replication_factor=1.unclean_abort=True.recovery=restart_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa3-696c-44bf-a14b-ef3904c18f47 | FLAKY | 1/2 |
| rptest.tests.partition_state_api_test.PartitionStateAPItest.test_local_summary | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa3-696b-42ef-89e9-2983e84523b7 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa6-c42c-42ca-bd89-59398e0acd42 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa6-c42d-443a-8763-3fbba47f6d50 | FLAKY | 1/3 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64007#0195efa6-c42d-4939-acb0-8054385add3a | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64147">build#64147</a></summary>


| test_id                                                                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64147#0195fa1c-78c7-4761-b7a2-96df93309cfd | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64147#0195fa1c-78c8-41e6-b17b-213aaae616a1 | FLAKY         | 1/2      |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move.replication_factor=3.unclean_abort=True.recovery=restart_recovery.compacted=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64147#0195fa1c-78c7-4761-b7a2-96df93309cfd | FLAKY         | 1/2      |
</details>
<!--/body-->

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -159,10 +164,24 @@ List all in json format.
+			if brokerDist {
+				sections.Add(secBrokerDist, func() { printBrokerDistribution(clusterPartitions) })
+			}
+			sections.Add(secClusterPartitions, func() { printClusterPartitions(f, clusterPartitions) })
```
We should add the conditional section to the `printClusterPartitions` function and accept `brokerDist` as a parameter.

Also, this command supports --format json/yaml. We should add this section by default to our formatter

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -83,6 +84,10 @@ List only the disabled partitions.
 List all in json format.
   rpk cluster partition list -a --format json
+
+Show how the partitions are distributed across brokers.
+  rpk cluster partitions list -a --broker-distribution
+  rpk cluster partitions foo bar --broker-distribution
```
I _think_ it's ok if we enable it by default. Showing the distribution does not require an additional request. 

What do you think?

daisukebe replied:
Sorry for the delay! Yea indeed it's a good idea.

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/status.go:
```diff
@@ -89,7 +90,24 @@ investigation. A few areas to investigate:
+				fmt.Printf("unable to query all partitions in the cluster: %v", err)
+			}
+
+			const (
+				secBalancerStatus = "Balancer status"
+				secBrokerDist     = "Broker distribution"
+			)
+			sections := out.NewSections(
+				out.ConditionalSectionHeaders(map[string]bool{
+					secBrokerDist:     true,
+					secBalancerStatus: true,
+				})...,
+			)
+			sections.Add(secBalancerStatus, func() { printBalancerStatus(status) })
```
Here I also think we can move the printing logic into the `printBalancerStatus` function

daisukebe commented:
Before updating this PR based on your feedback, wanted to have @r-vasquez to review this one https://github.com/redpanda-data/common-go/pull/50

daisukebe commented:
I've found I didn't need to edit common-go in this case. In the latest rebase, these are the changes:

* Print broker distribution out-of-box. No more `--broker-distribution` option
* Formatting is supported in the list command
* Moved all printing logic along with sections to `printBalancerStatus` in `balancer-status` and `printClusterPartitions` in `list`

kbatuigas commented on src/go/rpk/pkg/cli/cluster/partitions/status.go:
```diff
@@ -31,7 +32,8 @@ If continuous partition balancing is enabled, redpanda will continuously
 space than the configured limit.
 
-This command can be used to monitor the partition balancer status.
+This command can be used to monitor the partition balancer status as well as
```
```suggestion
Use this command to monitor the partition balancer status and
```

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -83,6 +88,10 @@ List only the disabled partitions.
+  rpk cluster partitions foo bar --broker-distribution
 `,
 		Run: func(cmd *cobra.Command, topics []string) {
 			f := p.Formatter
```
Now the help text should return `resp` instead of `[]rpadmin.ClusterPartition{}`

daisukebe replied:
Great catch! Thank you.

r-vasquez commented on src/go/rpk/pkg/cli/cluster/partitions/list.go:
```diff
@@ -30,6 +30,11 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+type response struct {
```
So, this is our first time with a breaking change in our `--format` response in rpk. We should avoid breaking current users where we can, but here, I think it's inevitable. This should prevent us from backporting to older branches.

It's fine, but we shouldn't backport and add this to the release notes. In retrospect, I shouldn't have returned an array in the first place, but instead an array as part of an object `{partitions:[]rpadmin.ClusterPartition}`

w.d.y.t @twmb

daisukebe replied:
It's ok to have this feature since 25.1.

daisukebe commented:
In the latest:

* Fixed help text per @kbatuigas 's suggestion
* Removed the examples with `--broker-distribution` option from the help text
* Fixed formatter's help text per @r-vasquez's suggestion

daisukebe commented:
In [ecdd1c5](https://github.com/redpanda-data/redpanda/pull/25197/commits/ecdd1c57eaaa88f24c0e85d0e835e39e000db620), I changed below:

* Renamed `Broker Distribution` to `Replica Distribution` per [advice](https://redpandadata.slack.com/archives/C01LK3UK2K1/p1742366507526459) from @JakeSCahill 
* Added `Leader Distribution` section to the list command

daisukebe commented:
Just rebased.

daisukebe commented:
@r-vasquez do you happen to have an idea on why these errors are arising when building?
```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/go/rpk/pkg/cli/cluster/partitions/list.go:34:25: undefined: LeaderDistribution
src/go/rpk/pkg/cli/cluster/partitions/list.go:35:25: undefined: ReplicaDistribution
src/go/rpk/pkg/cli/cluster/partitions/list.go:181:16: undefined: buildLeaderPerBroker
src/go/rpk/pkg/cli/cluster/partitions/list.go:182:17: undefined: buildReplicaPerBroker
src/go/rpk/pkg/cli/cluster/partitions/list.go:205:39: undefined: printLeaderDistribution
src/go/rpk/pkg/cli/cluster/partitions/list.go:206:40: undefined: printReplicaDistribution
src/go/rpk/pkg/cli/cluster/partitions/list.go:316:79: undefined: LeaderDistribution
src/go/rpk/pkg/cli/cluster/partitions/list.go:316:113: undefined: ReplicaDistribution
src/go/rpk/pkg/cli/cluster/partitions/status.go:134:18: undefined: buildReplicaPerBroker
src/go/rpk/pkg/cli/cluster/partitions/status.go:136:3: undefined: printReplicaDistribution
src/go/rpk/pkg/cli/cluster/partitions/status.go:136:3: too many errors
```

daisukebe commented:
/ci-repeat 3

daisukebe commented:
Hmm, it's still failing with the same reason. @r-vasquez thoughts?

r-vasquez commented:
@daisukebe It seems that you need to generate the bazel files, you can run `make bazel` in the root of the rpk repo for it

daisukebe commented:
Thanks @r-vasquez! Just pushed with this change. Let's see.

```diff
diff --git a/src/go/rpk/pkg/cli/cluster/partitions/BUILD b/src/go/rpk/pkg/cli/cluster/partitions/BUILD
index 032bc3d54f..b6780dd5d2 100644
--- a/src/go/rpk/pkg/cli/cluster/partitions/BUILD
+++ b/src/go/rpk/pkg/cli/cluster/partitions/BUILD
@@ -14,6 +14,7 @@ go_library(
         "toggle.go",
         "transfer_leadership.go",
         "unsafe_recover.go",
+        "util.go",
     ],
```

vbotbuildovich commented:
/backport v25.1.x
